node/network/statement-distribution/src/vstaging/tests/grid.rs

```diff
@@ -14,28 +14,2981 @@
+					);
+				}
+			);
+		}
+
+		// Receive Backed message.
+		overseer
+			.send(FromOrchestra::Communication {
+				msg: StatementDistributionMessage::Backed(candidate_hash),
+			})
+			.await;
+
+		// Should send an acknowledgement back to C.
+		{
+			assert_matches!(
+				overseer.recv().await,
+				AllMessages:: NetworkBridgeTx(
+					NetworkBridgeTxMessage::SendValidationMessage(
+						peers,
+						Versioned::VStaging(
+							protocol_vstaging::ValidationProtocol::StatementDistribution(
+								protocol_vstaging::StatementDistributionMessage::BackedCandidateKnown(ack),
+							),
+						),
+					)
+				) => {
+					assert_eq!(peers, vec![peer_c]);
+					assert_eq!(ack, BackedCandidateAcknowledgement {
+						candidate_hash,
+						statement_knowledge: StatementFilter {
+							seconded_in_group: bitvec::bitvec![u8, Lsb0; 0, 1, 1],
+							validated_in_group: bitvec::bitvec![u8, Lsb0; 0, 0, 0],
+						},
+					});
+				}
+			);
+
+			answer_expected_hypothetical_depth_request(&mut overseer, vec![], None, false).await;
+		}
+
+		// TODO: Receive a manifest about the same candidate from peer D. Contains different statements.
+		{
+			let manifest = BackedCandidateManifest {
+				relay_parent,
+				candidate_hash,
+				group_index: other_group,
+				para_id: other_para,
+				parent_head_data_hash: pvd.parent_head.hash(),
+				statement_knowledge: StatementFilter {
+					seconded_in_group: bitvec::bitvec![u8, Lsb0; 1, 1, 0],
+					validated_in_group: bitvec::bitvec![u8, Lsb0; 0, 0, 0],
+				},
+			};
+
+			send_peer_message(
+				&mut overseer,
+				peer_d.clone(),
+				protocol_vstaging::StatementDistributionMessage::BackedCandidateManifest(
+					manifest.clone(),
+				),
+			)
+			.await;
+
+			let expected_ack = BackedCandidateAcknowledgement {
+				candidate_hash,
+				statement_knowledge: StatementFilter {
+					seconded_in_group: bitvec::bitvec![u8, Lsb0; 0, 1, 1],
+					validated_in_group: bitvec::bitvec![u8, Lsb0; 0, 0, 0],
+				},
+			};
+
+			// Instantaneously acknowledge.
+			assert_matches!(
+				overseer.recv().await,
+				AllMessages:: NetworkBridgeTx(
+					NetworkBridgeTxMessage::SendValidationMessages(messages)
+				) => {
+					assert_eq!(messages.len(), 2);
+					assert_eq!(messages[0].0, vec![peer_d]);
+					assert_eq!(messages[1].0, vec![peer_d]);
+
+					assert_matches!(
+						&messages[0].1,
+						Versioned::VStaging(protocol_vstaging::ValidationProtocol::StatementDistribution(
+							protocol_vstaging::StatementDistributionMessage::BackedCandidateKnown(ack)
+						)) if *ack == expected_ack
+					);
+
+					assert_matches!(
+						&messages[1].1,
+						Versioned::VStaging(protocol_vstaging::ValidationProtocol::StatementDistribution(
+							protocol_vstaging::StatementDistributionMessage::Statement(r, s)
+						)) if *r == relay_parent && s.unchecked_payload() == &CompactStatement::Seconded(candidate_hash)
+					);
+				}
+			);
+		}
+
+		// TODO: Why is statement E not sent to peer D?
```

Comment:
@rphmeier Can you please take a look at this test? My understanding was that on manifest exchange we would send forward statements that the peer does not know about. Does the test need to be tweaked?

Reply:
but I'm confused - what's happening at L1363 if not that?

The local validator has statements [D, E], receives a manifest advertising [C, D], then responds with 2 messages:
  1. An acknowledgement (L1358)
  2. 1 statement (as D is known mutually). I think that another check at L1367 would confirm that the outgoing statement is statement E.

Reply:
(confirmed locally)

Reply:
Oh. Now I'm confused how I missed that. 😅 Thanks!